### PR TITLE
RepeatingCachedRootSource

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Created the RepeatingCachedRootSource which allows performance tests which are not dependent upon the read speed of the storage medium.

#### PR validation:

The added unit test passes.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: IOPool/Input/src/RepeatingCachedRootSource.cc [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+  auto input =
+      std::make_shared<InputFile>(physicalFileName.c_str(), "  Initiating request to open file ", InputType::Primary);
+  rootFile_ = makeRootFile(
+      logicalFileName, physicalFileName, 0 != nEventsToSkip, input, skipper, duplicateChecker, indexesIntoFiles);
+  rootFile_->reportOpened("repeating");
+
+  auto const& prodList = rootFile_->productRegistry()->productList();
+  productRegistryUpdate().updateFromInput(prodList);
+  //productRegistryUpdate().setFrozen();
```
Comment (makortel) [inline IOPool/Input/src/RepeatingCachedRootSource.cc:166]: Is this commented line useful?

File: IOPool/Input/src/RepeatingCachedRootSource.cc [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        }
+      }
+    }
+    rootFile_->rewind();
+  }
+}
+
+void RepeatingCachedRootSource::fillDescriptions(ConfigurationDescriptions& descriptions) {
+  ParameterSetDescription desc;
+  desc.setComment("Read only a few Events from EDM/Root files.");
```
Comment (makortel) [inline IOPool/Input/src/RepeatingCachedRootSource.cc:231]: I'd suggest to expand with something along
```suggestion
  desc.setComment("Read only a few Events from one EDM/Root file, and repeat them in sequence. The Events are required to be from the same Run and LuminosityBlock.");
```

File: IOPool/Input/src/RepeatingCachedRootSource.cc [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+  descriptions.add("source", desc);
+}
+
+// RepeatingCachedRootSource::RepeatingCachedRootSource(const RepeatingCachedRootSource& rhs)
+// {
+//    // do actual copying here;
+// }
+
+//RepeatingCachedRootSource::~RepeatingCachedRootSource() {
+//}
```
Comment (makortel) [inline IOPool/Input/src/RepeatingCachedRootSource.cc:248]: Are these commented fragments useful?

File: IOPool/Input/src/RepeatingCachedRootSource.cc [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    std::shared_ptr<WrapperBase> getProduct(unsigned int iStreamIndex,
+                                            BranchID const& k,
+                                            EDProductGetter const* ep) const;
+
+    class RCDelayedReader : public edm::DelayedReader {
+    public:
+      std::shared_ptr<edm::WrapperBase> getProduct_(edm::BranchID const& k, edm::EDProductGetter const* ep) final {
+        return m_source->getProduct(m_streamIndex, k, ep);
+      }
+      void mergeReaders_(edm::DelayedReader*) final { assert(false); }
```
Comment (makortel) [inline IOPool/Input/src/RepeatingCachedRootSource.cc:63]: Will this function be ever called?

Comment (Dr15Jones) [reply IOPool/Input/src/RepeatingCachedRootSource.cc:63]: This DelayedReader is only used on EventPrincipals and events do not do mergers. Therefore it should never be called.

File: IOPool/Input/src/RepeatingCachedRootSource.cc (hunk 1 above)
Comment (Dr15Jones) [reply IOPool/Input/src/RepeatingCachedRootSource.cc:166]: Nope.

Comment (Dr15Jones) [reply IOPool/Input/src/RepeatingCachedRootSource.cc]: removed

File: IOPool/Input/src/RepeatingCachedRootSource.cc (hunk 2 above)
Comment (Dr15Jones) [reply IOPool/Input/src/RepeatingCachedRootSource.cc]: done

File: IOPool/Input/src/RepeatingCachedRootSource.cc (hunk 3 above)
Comment (Dr15Jones) [reply IOPool/Input/src/RepeatingCachedRootSource.cc]: removed

**Comment by makortel** [comment]
@smuzaffar Could the code-checks failure be caused by #33330 (that introduced `ServiceWeakToken`) being merged but new IB not yet built, this PR touching a source file (`ProductResolvers.cc`) that uses `ServiceWeakToken` but nothing from the package where the `ServiceWeakToken` is defined (`FWCore/ServiceRegistry`), and the `FWCore/ServiceRegistry` is not checked out for the `code-checks`?

**Comment by smuzaffar** [comment]
yes looks like that is the reason. I am afraid you have to wait for next IB and then request for `code-checks`

**Comment by Dr15Jones** [comment]
code-checks

**Comment by Dr15Jones** [comment]
code-checks

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+1

**Comment by makortel** [comment]
@fwyzard This might be of interest to you as well (I recall you were asking about a Source that would do the real I/O prior to event loop and recycle the Events).

**Comment by fwyzard** [comment]
Thanks for the mention!
Is the decompression cached as well ?

**Comment by silviodonato** [comment]
+1

**Comment by Dr15Jones** [comment]
> Is the decompression cached as well ?

Yes

**Pull request merged**